oap-stdlib/src/main/java/oap/util/ConcurrentAssocMap.java

```diff
@@ -0,0 +1,93 @@
+package oap.util;
+
+
+import com.google.common.base.Preconditions;
+import lombok.NoArgsConstructor;
+
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * This class is a concurrent version of OAP AssocList<K,V>
+ * @param <K>
+ * @param <V>
+ */
+@SuppressWarnings( "AbstractClassName" )
+@NoArgsConstructor
+public abstract class ConcurrentAssocMap<K, V> {
+
+    private ConcurrentMap<K, V> entries = new ConcurrentHashMap<>();
+
+    protected abstract K keyOf( V label );
+
+    public ConcurrentAssocMap( ConcurrentMap<K, V> entries ) {
+        this.entries = entries;
+    }
+
+    public ConcurrentAssocMap( Collection<V> values ) {
+        entries = values.stream().collect( Collectors.toConcurrentMap( this::keyOf, Function.identity() ) );
```

Comment:
values.forEach( v -> this.entities.put(...) ?

Reply:
done